tests/aggregation/tests.py

```diff
@@ -1340,3 +1341,55 @@ def test_aggregation_random_ordering(self):
             ('Stuart Russell', 1),
             ('Peter Norvig', 2),
         ], lambda a: (a.name, a.contact_count), ordered=False)
+
+    def test_empty_result_optimization(self):
+        with self.assertNumQueries(0):
+            self.assertEqual(
+                Publisher.objects.none().aggregate(
+                    sum_awards=Sum('num_awards'),
+                )['sum_awards'], None
+            )
+        with self.assertNumQueries(0):
+            self.assertEqual(
+                Publisher.objects.none().aggregate(
+                    books_count=Count('book'),
+                )['books_count'], 0
+            )
+        raw_books_count = Func('book', function='COUNT')
+        raw_books_count.contains_aggregate = True
+        with self.assertNumQueries(1):
+            self.assertEqual(
+                Publisher.objects.none().aggregate(
+                    books_count=raw_books_count,
+                )['books_count'], 0
+            )
+
+    def test_coalesced_empty_result_set(self):
+        with self.assertNumQueries(0):
+            self.assertEqual(
+                Publisher.objects.none().aggregate(
+                    sum_awards=Coalesce(Sum('num_awards'), 0),
+                )['sum_awards'], 0
+            )
+        # Multiple expressions
+        with self.assertNumQueries(0):
+            self.assertEqual(
+                Publisher.objects.none().aggregate(
+                    sum_awards=Coalesce(Sum('num_awards'), None, 0),
+                )['sum_awards'], 0
+            )
+        # Nested coealesce
+        with self.assertNumQueries(0):
+            self.assertEqual(
+                Publisher.objects.none().aggregate(
+                    sum_awards=Coalesce(Coalesce(Sum('num_awards'), None), 0),
+                )['sum_awards'], 0
+            )
+        # Expression coalesce
+        self.assertIsInstance(
+            Store.objects.none().aggregate(
+                latest_opening=Coalesce(
+                    Max('original_opening'), RawSQL('CURRENT_TIMESTAMP', [])
```

Comment:
`Now()` could be used as well here, just wanted to make sure this test resisted the tests of time if decided of implementing `Now.get_empty_aggregate -> timezone.now()` even if it's technically not correct.